python/lsst/daf/butler/script/queryDatasets.py

```diff
@@ -78,22 +78,46 @@ def getAstropyTable(self, datasetTypeName):
             raise RuntimeError(f"No DatasetRefs were provided for dataset type {datasetTypeName}")
 
         refInfo = next(iter(self.datasetRefs))
+        dimensions = list(refInfo.datasetRef.dataId.full.keys())
         columnNames = ["type", "run", "id",
-                       *[str(item) for item in refInfo.datasetRef.dataId.keys()]]
+                       *[str(item) for item in dimensions]]
+
+        # Need to hint the column types for numbers since the per-row
+        # constructor of Table does not work this out on its own and sorting
+        # will not work properly without.
+        typeMap = {float: np.float, int: np.int64}
+        columnTypes = [None, None, np.int64,
+                       *[typeMap.get(type(value)) for value in refInfo.datasetRef.dataId.full.values()]]
         if refInfo.uri:
             columnNames.append("URI")
+            columnTypes.append(None)
 
         rows = []
-        for refInfo in sorted(self.datasetRefs):
+        for refInfo in self.datasetRefs:
             row = [datasetTypeName,
                    refInfo.datasetRef.run,
                    refInfo.datasetRef.id,
-                   *[str(value) for value in refInfo.datasetRef.dataId.values()]]
+                   *[value for value in refInfo.datasetRef.dataId.full.values()]]
             if refInfo.uri:
                 row.append(refInfo.uri)
             rows.append(row)
 
-        return AstropyTable(array(rows), names=columnNames)
+        dataset_table = AstropyTable(np.array(rows), names=columnNames, dtype=columnTypes)
+
+        # For sorting we want to ignore the id
+        # We also want to move temporal or spatial dimensions earlier
+        sort_first = ["type", "run"]
+        sort_early = []
+        sort_late = []
+        for dim in dimensions:
+            if dim.spatial or dim.temporal:
+                sort_early.append(str(dim))
```

Comment:
This will sort `visit` before `instrument`, which doesn't make sense, either.  Something like
```
sort_early.extend(dim.required.names)
```
instead of
```
sort_early.append(str(dim))
```
would solve that, but it'd lead to duplicates across the various `sort_` lists that you'd then need to remove.

Reply:
I can take another look. We haven't yet dealt with single runs having datasets with different instruments. One option is to simply state up front that instrument goes after run in the sorting and then leave everything else as is.

Reply:
Okay. Your fix worked fine.